drivers/misc/ethos_u/init.c

```diff
@@ -20,9 +21,27 @@ LOG_MODULE_REGISTER(ethos_u, CONFIG_ARM_ETHOS_U_LOG_LEVEL);
 
 /*******************************************************************************
  * Re-implementation/Overrides __((weak)) symbol functions from ethosu_driver.c
- * To handle mutex and semaphores
+ * To handle mutex, semaphores, and cache
  *******************************************************************************/
 
+void ethosu_invalidate_dcache(uint32_t *p, size_t bytes)
+{
+	if (p) {
+		sys_cache_data_invd_range(p, bytes);
+	} else {
+		sys_cache_data_invd_all();
```

Comment:
Not an expert in the ethosu driver itself, but calling `sys_cache_data_invd_all` from driver code looks completely wrong. According to documentation `sys_cache_data_invd_all` is indeed invalidate operation, not flush & invalidate.

Reply:
>Not an expert in the ethosu driver itself, but calling sys_cache_data_invd_all from driver code looks completely wrong.

That is what the ethos driver expects, if NULL is passed, then it expects the whole cache to be flushed or invalidated. https://github.com/zephyrproject-rtos/hal_ethos_u/blob/90ada2ea5681b2a2722a10d2898eac34c2510791/src/ethosu_driver.c#L127

> According to documentation sys_cache_data_invd_all is indeed invalidate operation, not flush & invalidate.

Yeah...., that's the intent there. To Invalidate, not flush & invalidate with the same function

Reply:
> Yeah...., that's the intent there.

But you can't just invalidate whole dcache in random place - it will eliminate all the data in dcache which are in modified state in comparison to lower layer of memory. I.e. you can eliminate recent writes to any variables, buffers, stack etc...

I'm just wondering how it's expected to work because it looks obviously wrong :)

Reply:
I understand your point... but this is following what ARM says to do...

_Under Data caching_
https://review.mlplatform.org/plugins/gitiles/ml/ethos-u/ethos-u-core-driver

Reply:
yeah, invalidating the whole cache before flushing it, it is definitely going to kill the kernel. As soon as you are popping for example the stack pointer, you are going to read garbage out of it.